File: Sources/Usage/Transport.swift
// POST transport over PlatformSupport's blocking HTTP client, plus the default
// client assembly.

import PlatformSupport
#if os(WASI)
import JavaScriptKit
#endif

/// The shared ingest endpoint. Every SDK reports to the same place, so it is not
/// part of the public API. A host may override it (tests/diagnostics) via
/// `hostProvidedIngestEndpoint()`.
private let defaultIngestEndpoint = "https://events.desertant.com/api/v1/ingest"
private var ingestEndpoint: String { hostProvidedIngestEndpoint() ?? defaultIngestEndpoint }

private let sendTimeoutSeconds = 5.0

/// Whether the key can ride an `Authorization` header on every path this
/// transport uses.
///
/// False on Android, whose host bridge takes a body and a content type only, and
/// on wasm, where a browser's unload flush is a `sendBeacon` and cannot carry a
/// header; one wasm binary is the same code for a page and for a Node process, so
/// it keeps one answer for both rather than branching on the host. Those two keep
/// the key in the body.
private var keyRidesInHeader: Bool {
    if !httpSupportsRequestHeaders { return false }
    #if os(WASI)
    return false
    #else
    return true
    #endif
}

/// A `send` transport that POSTs the serialized body to `endpoint`.
///
/// The HTTP client is async, so every flush is dispatched fire-and-forget on a
/// detached task. `beacon` only changes the path on wasm, where it uses
/// `navigator.sendBeacon`.
///
/// Sends nothing while `usageDisabled()` is on, read per send: the last guard
/// behind the client's own (`ClientDeps.disabled`), for a host that pairs this
/// transport with a client of its own.
public func makeSend(endpoint: String, bearerKey: String? = nil) -> @Sendable (IngestBody, SendOptions) -> Void {
    makeSend(endpoint: endpoint, bearerKey: bearerKey, registry: .shared)
}

/// `registry` is a seam for tests: a send must be in it by the time the
/// returned closure returns, which only a private registry lets a test observe.
/// `disabled` is one too, since the suites run with the switch on.
func makeSend(
    endpoint: String,
    bearerKey: String? = nil,
    registry: InflightSends,
    disabled: @escaping @Sendable () -> Bool = usageDisabled
) -> @Sendable (IngestBody, SendOptions) -> Void {
    { body, opts in
        if disabled() { return }
        // Best-effort: a body we cannot serialize is dropped rather than thrown
        // (the transport is fire-and-forget). These types always encode.
        guard let json = try? buildBody(body) else { return }
        let payload = Array(json.utf8)
        let debug = telemetryDebugEnabled()
        var headers: [String: String] = [:]
        if let bearerKey, keyRidesInHeader { headers["Authorization"] = "Bearer \(bearerKey)" }
        if debug { print("[usage] POST \(endpoint)\n[usage] body: \(json)") }
        #if os(WASI)
        // On the browser, an unload flush must use navigator.sendBeacon (a normal
        // fetch is cancelled as the page goes away). text/plain keeps it a CORS
        // "simple" request (the server parses the body as JSON regardless).
        if opts.beacon, jsSendBeacon(endpoint, payload) { return }
        #endif
        // Registered before this returns, so a caller's `flushTelemetry()` awaits it.
        dispatchTrackedSend(into: registry) { [headers] in
            do {
                // Bounded like the Node and Kotlin ports' sends: `flushTelemetry()`
                // awaits this, and a blackholed endpoint would otherwise hold the
                // caller for URLSession's 60 s default.
                let response = try await httpPOST(
                    endpoint, body: payload, contentType: "application/json", headers: headers,
                    timeout: sendTimeoutSeconds
                )
                if debug {
                    let text = String(decoding: response.body, as: UTF8.self)
                    print("[usage] response: \(response.status) \(text)")
                }
            } catch {
                if debug { print("[usage] send failed: \(error)") }
            }
        }
    }
}

#if os(WASI)
/// `navigator.sendBeacon(endpoint, Blob([payload], {type: text/plain}))`.
private func jsSendBeacon(_ url: String, _ payload: [UInt8]) -> Bool {
    guard let navigator = JSObject.global.navigator.object,
          let sendBeacon = navigator.sendBeacon.function,
          let blobType = JSObject.global.Blob.function else { return false }
    let parts = JSObject.global.Array.function!.new()
    _ = parts.push!(JSTypedArray<UInt8>(payload).jsValue)
    let options = JSObject.global.Object.function!.new()
    options.type = "text/plain;charset=UTF-8".jsValue
    let blob = blobType.new(parts.jsValue, options.jsValue)
    // `this: navigator` is required: sendBeacon is a Navigator method and throws
    // "Illegal invocation" when called detached.
    return sendBeacon(this: navigator, url.jsValue, blob.jsValue).boolean ?? false
}
#endif

/// Build a client wired to the shared endpoint, the system clock, a POST
/// transport, and platform-native storage. Everything is derived and persisted
/// internally: attribution is the app's platform identity (bundle id on Apple,
/// package name on Android, hostname on web; see `defaultAppIdentifier`), sent
/// as `app.id`; the device id + re-emit state live in the platform store.
///
/// - Parameters:
///   - appId: overrides the auto-derived app identity. Falls back to the host
///     override (`globalThis.__dalAppId` on WASI, `DAL_APP_ID` env var natively)
///     and then the platform default. Sent as `app.id`; namespaces persisted state.
///   - key: a publishable API key, if the host has one (usually nil; native
///     attributes by `app.id`, browsers by Origin). Falls back to the host
///     override (`globalThis.__dalApiKey` on WASI, `DAL_API_KEY` env var natively).
///     Sent as `Authorization: Bearer` where the transport sets headers, and in
///     the body where it cannot (`keyRidesInHeader`).
///   - deviceId: overrides the device id. Defaults to a host-provided id
///     (`globalThis.__dalDeviceId`, for server-side Node) or the generated,
///     persisted per-install UUID.
///   - storage: overrides the persistence backend (e.g. tests).
///   - context: overrides the per-event `context` provider. Defaults to the
///     cached host facts (`DeviceContext`), cut to the server set when `platform`
///     is "server" or the device id was supplied rather than generated here.
///     Whatever it returns is sanitized before it is sent (`sanitizeContext`),
///     and ignored while the context opt-out is on (`DesertAnt.sendsDeviceContext`,
///     `DAL_USAGE_CONTEXT_DISABLED`).
///   - send: overrides the transport. Defaults to the real POST; a caller-supplied
///     one wins (tests), which is how a test reads the platform tag and the key's
///     placement that this function decides.
///   - disabled: the usage opt-out the client reads before every step. Defaults
///     to `usageDisabled`; a test passes its own, since the suites run with the
///     switch on.
public func makeClient(
    appId: String? = nil,
    key: String? = nil,
    sdk: SDKInfo = SDKInfo(),
    deviceId: String? = nil,
    platform: String = defaultPlatform,
    windowMs: Int64 = dayMs,
    emitIntervalMs: Int64? = nil,
    callCount: (() -> Int)? = nil,
    context: (() -> [String: String]?)? = nil,
    storage: UsageStorage? = nil,
    send: ((IngestBody, SendOptions) -> Void)? = nil,
    disabled: @escaping () -> Bool = usageDisabled
) -> UsageClient {
    let resolvedAppId = appId ?? hostProvidedAppId() ?? defaultAppIdentifier()
    let resolvedKey = trimmedKey(key) ?? hostProvidedApiKey()
    let namespace = resolvedKey ?? resolvedAppId    // state namespaced per attribution identity
    let store = storage ?? defaultStorage()
    let device = resolveDeviceId(deviceId, store)
    // An id from the caller (a tenant's) or the host is not this machine's to
    // describe. Inference passes the persisted id explicitly on its default path,
    // which is why an explicit id equal to it still counts as generated.
    let suppliedId = deviceId ?? hostProvidedDeviceId()
    let deviceIdSupplied = suppliedId != nil && suppliedId != store.get(deviceIdKey)
    // Coalesce a continuously-running server's delta loads to hourly by default.
    let resolvedEmitInterval = emitIntervalMs ?? (platform == "server" ? hourMs : 0)
    return UsageClient(ClientDeps(
        deviceId: device,
        key: resolvedKey,
        keyInBody: !keyRidesInHeader,
        appId: resolvedAppId,
        sdk: sdk,
        platform: platform,
        callCount: callCount,
        context: context ?? defaultContextProvider(platform: platform, deviceIdSupplied: deviceIdSupplied),
        windowMs: windowMs,
        emitIntervalMs: resolvedEmitInterval,
        now: systemNowMs,
        loadState: { store.loadState(namespace, device) },
        saveState: { store.saveState($0, namespace, device) },
        // The key is only known here, so the real transport is built here too; a
        // caller-supplied one still wins (tests).
        send: send ?? makeSend(endpoint: ingestEndpoint, bearerKey: resolvedKey),
        disabled: disabled
    ))
}

File: packages/tongue-kotlin/src/main/kotlin/ai/desertant/tongue/usage/Transport.kt
package ai.desertant.tongue.usage

import java.net.HttpURLConnection
import java.net.URL
import java.util.Locale
import java.util.TimeZone
import java.util.concurrent.Executors
import java.util.concurrent.ThreadFactory

/**
 * Time, identity and the POST transport — the pieces core keeps in
 * `Identity.swift` and `Transport.swift`.
 *
 * `HttpURLConnection` rather than a client library: the jar must stay
 * free of declared dependencies, and it exists on every JVM and Android level this SDK
 * supports.
 */

/** The shared ingest endpoint. Every SDK reports to the same place. */
internal const val INGEST_ENDPOINT: String = "https://events.desertant.com/api/v1/ingest"

/**
 * The ingest endpoint for this process. Core lets a host override it for tests
 * and local capture (`hostProvidedIngestEndpoint`), and this reads the same
 * variable; the system property is the same override for a caller that cannot
 * set the environment.
 */
internal fun ingestEndpoint(): String =
    setting("DAL_INGEST_ENDPOINT") ?: INGEST_ENDPOINT

/**
 * The publishable API key for this process, or null. A key set in code
 * (`DesertAnt.apiKey`) wins, matching core's `hostProvidedApiKey()`; otherwise
 * it is read from the environment as core does, then from the same-named system
 * property. Trimmed: a key read from a secret file often ends in a newline, which
 * the body tolerated but `setRequestProperty` rejects, dropping the POST.
 */
internal fun apiKey(): String? =
    ai.desertant.tongue.DesertAnt.apiKey?.trim()?.takeIf { it.isNotEmpty() }
        ?: setting("DAL_API_KEY")?.trim()?.takeIf { it.isNotEmpty() }

/**
 * A host-provided device id, or null. Core's `hostProvidedDeviceId()` reads the
 * same variable, and when set it replaces the persisted per-install id: a server
 * that knows its own device identity names it rather than being counted anew
 * per process.
 */
internal fun hostProvidedDeviceId(): String? = setting("DAL_DEVICE_ID")

/**
 * How this port reads the process environment. A seam for tests only: the Gradle
 * test task sets `DAL_USAGE_DISABLED` in the environment, and the environment
 * wins, so without it the system property path could never be exercised.
 */
internal var readEnvironment: (String) -> String? = System::getenv

/** An environment variable, then the same-named system property, then null. */
internal fun setting(name: String): String? =
    readEnvironment(name)?.takeIf { it.isNotEmpty() }
        ?: System.getProperty(name)?.takeIf { it.isNotEmpty() }

/**
 * Format epoch milliseconds as `2024-01-02T03:04:05.678Z`.
 *
 * Hand-formatted from a UTC calendar rather than `java.time`: `Instant` and
 * `DateTimeFormatter` need API 26 without desugaring, and this SDK supports
 * older Android. Matches core's `iso8601`, which hand-formats for the same
 * reason (no Foundation on its Android target).
 */
internal fun iso8601(epochMs: Long): String {
    val calendar = java.util.Calendar.getInstance(TimeZone.getTimeZone("UTC"), Locale.US)
    calendar.timeInMillis = epochMs
    return String.format(
        Locale.US,
        "%04d-%02d-%02dT%02d:%02d:%02d.%03dZ",
        calendar.get(java.util.Calendar.YEAR),
        calendar.get(java.util.Calendar.MONTH) + 1,
        calendar.get(java.util.Calendar.DAY_OF_MONTH),
        calendar.get(java.util.Calendar.HOUR_OF_DAY),
        calendar.get(java.util.Calendar.MINUTE),
        calendar.get(java.util.Calendar.SECOND),
        calendar.get(java.util.Calendar.MILLISECOND),
    )
}

/**
 * One daemon thread, shared. A send must never block a detection and must never
 * keep a JVM alive: a short-lived CLI that detects once should exit immediately,
 * not linger on a non-daemon pool.
 */
private val sender = Executors.newSingleThreadExecutor(
    ThreadFactory { runnable ->
        Thread(runnable, "tongue-usage").apply { isDaemon = true; priority = Thread.MIN_PRIORITY }
    },
)

/**
 * The platform tag this build reports. The endpoint accepts exactly
 * ios|android|web|server and rejects anything else with a 400, which drops the
 * event silently: a JVM is a `server`, not a `jvm`.
 */
internal fun defaultPlatform(): String = if (isAndroid()) "android" else "server"

/**
 * A send in flight. `await` blocks until the POST has finished or `deadlineNanos`
 * (a `System.nanoTime()` value) passes, so a short-lived caller can be sure the
 * request left the process before it exits. A send still running at the deadline
 * is left to finish on the sender: it may yet land.
 */
internal fun interface SendHandle {
    fun await(deadlineNanos: Long)
}

/**
 * How long a flush waits for its sends, all of them together. The connection's
 * own timeouts bound one send to about 10 s, and a flush may await two, which on
 * Android's main thread is an ANR; the Node port bounds each send to 5 s too.
 */
internal const val FLUSH_WAIT_MS: Long = 5_000L

/** Await with a deadline `FLUSH_WAIT_MS` from now. */
internal fun SendHandle.await() = await(System.nanoTime() + FLUSH_WAIT_MS * 1_000_000L)

/** Await every handle against one shared deadline, so two sends cannot wait 2x. */
internal fun List<SendHandle>.awaitAll() {
    val deadline = System.nanoTime() + FLUSH_WAIT_MS * 1_000_000L
    forEach { it.await(deadline) }
}

/**
 * A `send` that POSTs the serialized body, fire and forget unless awaited.
 *
 * The key rides an `Authorization` header, not the body: every transport this
 * port runs on sets request headers (Android's `HttpURLConnection` included,
 * unlike core's Android host bridge), and the endpoint prefers the header.
 */
internal fun makeSend(endpoint: String = INGEST_ENDPOINT, bearerKey: String? = null): (IngestBody) -> SendHandle? = { body ->
    val json = runCatching { buildBody(body) }.getOrNull()
    if (json == null) {
        null
    } else {
        runCatching {
            sender.submit {
                runCatching {
                    val connection = URL(endpoint).openConnection() as HttpURLConnection
                    connection.requestMethod = "POST"
                    connection.doOutput = true
                    connection.connectTimeout = 5_000
                    connection.readTimeout = 5_000
                    connection.setRequestProperty("Content-Type", "application/json")
                    if (bearerKey != null) {
                        connection.setRequestProperty("Authorization", "Bearer $bearerKey")
                    }
                    connection.outputStream.use { it.write(json.toByteArray(Charsets.UTF_8)) }
                    connection.responseCode // the request is not sent until this is read
                    connection.disconnect()
                }
            }
        }.getOrNull()?.let { future -> SendHandle { deadline -> awaitFuture(future, deadline) } }
    }
}

/** Wait for the POST until the deadline. A cancellation request is re-flagged rather than swallowed. */
private fun awaitFuture(future: java.util.concurrent.Future<*>, deadlineNanos: Long) {
    try {
        future.get(maxOf(0L, deadlineNanos - System.nanoTime()), java.util.concurrent.TimeUnit.NANOSECONDS)
    } catch (_: java.util.concurrent.TimeoutException) {
        // Not cancelled: the POST keeps running on the sender and may still land.
    } catch (interrupted: InterruptedException) {
        Thread.currentThread().interrupt()
    } catch (_: java.util.concurrent.ExecutionException) {
        // The body already swallows its own failures; this cannot surface one.
    }
}

/**
 * The application identity used for keyless attribution: core sends the bundle id
 * on Apple and the package name on Android. Reflectively on Android (see
 * Storage.kt for why), otherwise the main class or process name.
 */
internal fun defaultAppIdentifier(context: Any? = null): String {
    if (context != null) {
        val packageName = runCatching {
            context.javaClass.getMethod("getPackageName").invoke(context) as? String
        }.getOrNull()
        if (!packageName.isNullOrEmpty()) return packageName
    }
    System.getenv("DAL_APP_ID")?.takeIf { it.isNotEmpty() }?.let { return it }
    return System.getProperty("java.vm.name")?.takeIf { it.isNotEmpty() } ?: "unknown"
}

/**
 * Whether usage reporting is switched off, right now: `DesertAnt.usageDisabled`
 * set in code, or `DAL_USAGE_DISABLED` in the environment or as a system
 * property, under `flagIsSet`. Read per call, so an app can hold it on until its
 * user consents and clear it then. See packages/tongue-node/USAGE.md.
 */
internal fun usageDisabled(): Boolean =
    ai.desertant.tongue.DesertAnt.usageDisabled ||
        flagIsSet(readEnvironment("DAL_USAGE_DISABLED")) ||
        flagIsSet(System.getProperty("DAL_USAGE_DISABLED"))

/**
 * Build a client wired to the shared endpoint, the system clock, a POST transport,
 * the best available storage and the default device context. Mirrors core's
 * `makeClient`.
 */
internal fun makeClient(
    context: Any? = null,
    sdkVersion: String,
    storage: UsageStorage = defaultStorage(context),
    send: ((IngestBody) -> SendHandle?)? = null,
    now: () -> Long = System::currentTimeMillis,
): UsageClient {
    val appId = defaultAppIdentifier(context)
    val key = apiKey()
    val namespace = key ?: appId
    val hostDevice = hostProvidedDeviceId()
    // A host id equal to the stored one is still this device's own, as core
    // and the Node port count it.
    val stored = if (hostDevice != null) storage.storedDeviceId() else null
    val device = hostDevice ?: storage.persistentDeviceId()
    val platform = defaultPlatform()
    return UsageClient(
        ClientDeps(
            deviceId = device,
            appId = appId,
            platform = platform,
            sdkVersion = sdkVersion,
            context = defaultContextProvider(
                platform,
                deviceIdSupplied = hostDevice != null && hostDevice != stored,
                facts = runCatching { detectDeviceFacts(context) }.getOrDefault(DeviceFacts()),
            ),
            now = now,
            loadState = { storage.loadState(namespace, device) },
            saveState = { storage.saveState(it, namespace, device) },
            // The key is only known here, so the real transport is built here too;
            // a caller-supplied one still wins (tests).
            send = send ?: makeSend(ingestEndpoint(), key),
        ),
    )
}
